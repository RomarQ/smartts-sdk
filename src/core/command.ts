import type { IToString } from '../typings';

import Utils, { LineInfo } from '../misc/utils';

class C_DefineLocal implements IToString {
    constructor(public name: string, public value: unknown, public mutable: boolean, public line = new LineInfo()) {}

    toString() {
        return `(defineLocal "${this.name}" ${this.value} ${Utils.capitalizeBoolean(this.mutable)} ${this.line})`;
    }
}
export const DefineVar = (name: string, value: unknown, mutable = true, line = new LineInfo()) =>
    new C_DefineLocal(name, value, mutable, line);

class C_Verify implements IToString {
    constructor(public condition: IToString, public errorMsg: IToString, public line = new LineInfo()) {}

    toString() {
        return `(verify ${this.condition} ${this.errorMsg} ${this.line})`;
    }
}
export const Require = (condition: IToString, errorMsg: IToString, line = new LineInfo()) =>
    new C_Verify(condition, errorMsg, line);

class C_SetValue implements IToString {
    name = 'set';
    constructor(public source: unknown, public value: unknown, public line = new LineInfo()) {}

    toString() {
        return `(${this.name} ${this.source} ${this.value} ${this.line})`;
    }
}
export const SetValue = (source: unknown, value: unknown, line = new LineInfo()) => new C_SetValue(source, value, line);

const Commands = {
    DefineVar,
    Require,
    SetValue,
};

export default Commands;
