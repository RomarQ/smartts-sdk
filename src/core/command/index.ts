export * from './controls';
export * from './variables';
export * from './function';
