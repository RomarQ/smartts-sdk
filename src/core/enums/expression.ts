export enum ExpressionAtom {
    // Math
    add = 'add',
    mul = 'mul',
    sub = 'sub',
    truediv = 'truediv',
    // Comparison
    eq = 'eq',
    neq = 'neq',
    lt = 'lt',
    gt = 'gt',
    le = 'le',
    ge = 'ge',
    //
    items = 'items',
    update_map = 'update_map',
    getItem = 'getItem',
    getItemDefault = 'getItemDefault',
    getItemMessage = 'getItemMessage',
    attr = 'attr',
    params = 'params',
    getLocal = 'getLocal',
    iter = 'iter',
    // Blockchain property
    amount = 'amount',
    balance = 'balance',
    chain_id = 'chain_id',
    level = 'level',
    now = 'now',
    self = 'self',
    self_address = 'self_address',
    sender = 'sender',
    source = 'source',
    total_voting_power = 'total_voting_power',
    voting_power = 'voting_power',
    // Operation
    operations = 'operations',
    transfer = 'transfer',
    set_delegate = 'set_delegate',
    create_contract = 'create_contract',
    // List instructions
    cons = 'cons',
    //
    contract = 'contract',
    // Variant
    openVariant = 'openVariant',
    isVariant = 'isVariant',
    cases_arg = 'cases_arg',
    variant_arg = 'variant_arg',
}

export default ExpressionAtom;
