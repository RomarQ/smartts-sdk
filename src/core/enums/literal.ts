export enum LiteralAtom {
    record = 'record',
    variant = 'variant',
    chain_id_cst = 'chain_id_cst',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    Some = 'Some',
    None = 'None',
    Left = 'Left',
    Right = 'Right',
    map = 'map',
    big_map = 'big_map',
    signature = 'signature',
    tuple = 'tuple',
    ticket = 'ticket',
    sapling_state = 'sapling_empty_state',
}

export default LiteralAtom;
