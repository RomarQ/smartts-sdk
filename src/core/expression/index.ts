export * from './literal';
export * from './expression';
export * from './logic';
