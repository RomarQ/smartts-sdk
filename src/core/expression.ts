import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { GetProperty, LambdaArgument } from '../expression/variables';
import { capitalizeBoolean, LineInfo, parenthesis } from '../misc/utils';
import ValueAtom from './enums/literal';
import TypeAtom from './enums/type';

export class Expression<T extends ValueAtom> implements IExpression<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;
    private args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return GetProperty(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class LiteralExpression<T extends ValueAtom> implements ILiteral<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;

    constructor(private name: T, private values: (number | string | boolean | IExpression)[], private line: LineInfo) {}

    toString() {
        switch (this.name) {
            case ValueAtom.unit:
                return parenthesis(this.name);
            case ValueAtom.nat:
            case ValueAtom.int:
            case ValueAtom.mutez:
            case ValueAtom.bytes:
            case ValueAtom.bool:
            case ValueAtom.string:
            case ValueAtom.chain_id_cst:
            case ValueAtom.timestamp:
            case ValueAtom.bls12_381_fr:
            case ValueAtom.bls12_381_g1:
            case ValueAtom.bls12_381_g2:
            case ValueAtom.key:
            case ValueAtom.key_hash:
            case ValueAtom.signature:
            case ValueAtom.address:
                return `(literal (${this.name} ${this.values.join(' ')}) ${this.line})`;
            case ValueAtom.list:
            case ValueAtom.set:
            case ValueAtom.tuple:
                return `(${this.name} ${this.line} ${this.values.join(' ')})`;
            case ValueAtom.Some:
            case ValueAtom.None:
                return `(${ValueAtom.variant} "${this.name}" ${this.values.join(' ') || `(${ValueAtom.unit})`} ${
                    this.line
                })`;
            default:
                return `(${this.name} ${this.values.join(' ')} ${this.line})`;
        }
    }
}

export class RecordLiteral implements ILiteral<ValueAtom.record> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as ValueAtom.record;

    constructor(private fields: Record<string, IExpression>, private line: LineInfo) {}

    private buildFields = (fields: Record<string, IExpression>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(${TypeAtom.record} ${this.line} ${this.buildFields(this.fields).join(' ')})`;
    }
}

export class MapLiteral<T extends ValueAtom.map | ValueAtom.big_map> implements ILiteral<T> {
    _isExpression = true as const;
    _type = {} as T;
    type = {} as IType;

    constructor(
        private prim: T,
        private rows: IExpression[][],
        keyType: IType,
        valueType: IType,
        private line: LineInfo,
    ) {}

    private buildEntry = ([key, value]: IExpression[]) => {
        return `(${key.toString()} ${value.toString()})`;
    };

    toString() {
        return `(${this.prim} ${this.line} ${this.rows.map(this.buildEntry).join(' ')})`;
    }
}

export class LambdaLiteral implements ILiteral<ValueAtom.lambda> {
    _isExpression = true as const;
    // Used for type checking
    _type = ValueAtom.lambda as const;
    type = {} as unknown as IType;
    static idCounter = 0;
    private identifier: number;
    private withStorage?: 'read-write' | 'read-only';
    private withOperations = false;

    constructor(
        private statements: IStatement[],
        private inType: IType,
        private argumentName: string,
        private line: LineInfo,
    ) {
        this.identifier = ++LambdaLiteral.idCounter;
    }

    public get id(): Readonly<number> {
        return this.identifier;
    }

    public setInputType(type: IType) {
        this.inType = type;
        return this;
    }

    public code(buildStatements: (arg: IExpression) => IStatement[]) {
        const param = LambdaArgument(this.argumentName, this.inType, this.id, this.line);
        this.statements = buildStatements(param);
        return this;
    }

    toString() {
        return `(${ValueAtom.lambda} ${this.id} ${this.withStorage || 'None'} ${capitalizeBoolean(
            this.withOperations,
        )} "${this.argumentName}" ${this.line} (${this.statements.join(' ')}))`;
    }
}
