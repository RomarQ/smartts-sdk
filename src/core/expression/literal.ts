import type { IType } from '../../typings/type';
import type { ILiteral } from '../../typings/literal';

import { capitalizeBoolean, LineInfo, parenthesis, quote } from '../../misc/utils';
import { TUnknown, TRecord, TMap, TBig_map } from '../type';
import { IExpression, IExpressionKind } from '../../typings/expression';
import { Layout } from '../enums/layout';
import LiteralAtom from '../enums/literal';
import TypeAtom from '../enums/type';
import { IStatement } from '../../typings/statement';
import { Expression } from '.';

class LiteralExpression<T extends TypeAtom> implements ILiteral<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = null as unknown as T;
    type = {} as unknown as IType;

    constructor(
        private name: LiteralAtom,
        private values: (number | string | boolean | IExpression)[],
        private line: LineInfo,
    ) {}

    toString() {
        switch (this.name) {
            case LiteralAtom.unit:
                return parenthesis(this.name);
            case LiteralAtom.nat:
            case LiteralAtom.int:
            case LiteralAtom.mutez:
            case LiteralAtom.bytes:
            case LiteralAtom.bool:
            case LiteralAtom.string:
            case LiteralAtom.chain_id_cst:
            case LiteralAtom.timestamp:
            case LiteralAtom.bls12_381_fr:
            case LiteralAtom.bls12_381_g1:
            case LiteralAtom.bls12_381_g2:
            case LiteralAtom.key:
            case LiteralAtom.key_hash:
            case LiteralAtom.signature:
            case LiteralAtom.address:
                return `(literal (${this.name} ${this.values.join(' ')}) ${this.line})`;
            case LiteralAtom.list:
            case LiteralAtom.set:
            case LiteralAtom.tuple:
                return `(${this.name} ${this.line} ${this.values.join(' ')})`;
            case LiteralAtom.Some:
            case LiteralAtom.None:
                return `(${LiteralAtom.variant} "${this.name}" ${this.values.join(' ') || `(${LiteralAtom.unit})`} ${
                    this.line
                })`;
            default:
                return `(${this.name} ${this.values.join(' ')} ${this.line})`;
        }
    }
}

class RecordLiteral implements ILiteral<TypeAtom.record> {
    _isExpression = true as const;
    _type = TypeAtom.record as const;

    type: IType;

    constructor(private fields: Record<string, ILiteral<unknown>>, private line: LineInfo) {
        // Compute the record type (use rightcombs by default)
        this.type = TRecord(
            Object.entries(fields).reduce(
                (pv, [field, expr]) => ({
                    ...pv,
                    [field]: expr.type.toString(),
                }),
                {},
            ),
            Layout.right_comb,
        );
    }

    private buildFields = (fields: Record<string, IExpressionKind>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(${TypeAtom.record} ${this.line} ${this.buildFields(this.fields).join(' ')})`;
    }

    toType() {
        return this.type.toString();
    }
}

class MapLiteral<T extends TypeAtom.map | TypeAtom.big_map> implements ILiteral<T> {
    _isExpression = true as const;
    _type: T;

    type: IType;

    constructor(
        private prim: LiteralAtom.map | LiteralAtom.big_map,
        private rows: IExpression[][],
        keyType: IType,
        valueType: IType,
        private line: LineInfo,
    ) {
        if (prim === LiteralAtom.map) {
            this.type = TMap(keyType, valueType);
        } else {
            this.type = TBig_map(keyType, valueType);
        }
        // Just for typing purposes
        this._type = null as unknown as T;
    }

    private buildEntry = ([key, value]: IExpression[]) => {
        return `(${key.toString()} ${value.toString()})`;
    };

    toString() {
        return `(${this.prim} ${this.line} ${this.rows.map(this.buildEntry).join(' ')})`;
    }

    toType() {
        return this.type.toString();
    }
}

class LambdaLiteral implements ILiteral<TypeAtom.lambda> {
    _isExpression = true as const;
    // Used for type checking
    _type = TypeAtom.lambda as const;
    type = {} as unknown as IType;
    static idCounter = 0;
    private id: number;
    private withStorage = false;
    private withOperations = false;
    private statements: IStatement[] = [];

    constructor(private inType: IType, private line: LineInfo) {
        this.id = ++LambdaLiteral.idCounter;
    }

    public inputType(type: IType) {
        this.inType = type;
        return this;
    }

    public code(buildStatements: (arg: IExpression) => IStatement[]) {
        const param = new Expression('lambdaParams', `${this.id}`, '"arg"', new LineInfo(), this.inType);
        this.statements = buildStatements(param);
        return this;
    }

    toString() {
        return `(lambda ${this.id} ${capitalizeBoolean(this.withStorage)} ${capitalizeBoolean(
            this.withOperations,
        )} "arg" ${this.line} (${this.statements.join(' ')}))`;
    }
}

// Singletons
export const Unit = (line = new LineInfo()) => new LiteralExpression<TypeAtom.unit>(LiteralAtom.unit, [], line);
export const Nat = (value: number, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.nat>(LiteralAtom.nat, [value], line);
export const Int = (value: number, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.int>(LiteralAtom.int, [value], line);
export const Mutez = (value: number, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.mutez>(LiteralAtom.mutez, [value], line);
export const String = (value: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.string>(LiteralAtom.string, [quote(value)], line);
export const Bool = (value: boolean, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.bool>(LiteralAtom.bool, [capitalizeBoolean(value)], line);
export const Address = (address: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.address>(LiteralAtom.address, [address], line);
export const Timestamp = (timestamp: number, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.timestamp>(LiteralAtom.timestamp, [timestamp], line);
export const Chain_id = (chainID: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.chain_id>(LiteralAtom.chain_id_cst, [chainID], line);
export const Bytes = (bytes: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.bytes>(LiteralAtom.bytes, [bytes], line);
export const Bls12_381_fr = (fr: string | number, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.bls12_381_fr>(LiteralAtom.bls12_381_fr, [fr], line);
export const Bls12_381_g1 = (bytes: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.bls12_381_g1>(LiteralAtom.bls12_381_g1, [bytes], line);
export const Bls12_381_g2 = (bytes: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.bls12_381_g2>(LiteralAtom.bls12_381_g2, [bytes], line);
export const Key = (key: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.key>(LiteralAtom.key, [key], line);
export const Key_hash = (key_hash: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.key_hash>(LiteralAtom.key_hash, [key_hash], line);
export const Signature = (signature: string, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.signature>(LiteralAtom.signature, [signature], line);
// Containers
export const List = (items: IExpression[], line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.list>(LiteralAtom.list, items, line);
export const Set = (items: IExpression[], line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.set>(LiteralAtom.set, items, line);
export const Some = (value: IExpressionKind, line = new LineInfo()) =>
    new LiteralExpression(LiteralAtom.Some, [value], line);
export const None = (line = new LineInfo()) => new LiteralExpression(LiteralAtom.None, [], line);
export const Map = (
    rows: IExpression[][] = [],
    keyType: IType = TUnknown,
    valueType: IType = TUnknown,
    line = new LineInfo(),
) => new MapLiteral<TypeAtom.map>(LiteralAtom.map, rows, keyType, valueType, line);
export const Big_map = (
    rows: IExpression[][] = [],
    keyType: IType = TUnknown,
    valueType: IType = TUnknown,
    line = new LineInfo(),
) => new MapLiteral<TypeAtom.big_map>(LiteralAtom.big_map, rows, keyType, valueType, line);
export const Pair = (left: IExpression, right: IExpression, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.tuple>(LiteralAtom.tuple, [left, right], line);
export const Ticket = (content: IExpression, amount: LiteralExpression<TypeAtom.nat>, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.ticket>(LiteralAtom.ticket, [content, amount], line);
export const Sapling_state = (memo: number, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.sapling_state>(LiteralAtom.sapling_state, [memo], line);
export const Lambda = (inType: IType = TUnknown, line = new LineInfo()) => new LambdaLiteral(inType, line);
export const Record = (fields: Record<string, ILiteral<unknown>>, line = new LineInfo()) =>
    new RecordLiteral(fields, line);
export const Variant = (field: string, value: IExpression, line = new LineInfo()) =>
    new LiteralExpression<TypeAtom.variant>(LiteralAtom.variant, [field, value], line);

const Literals = {
    // Singletons
    Unit,
    Nat,
    Int,
    Mutez,
    String,
    Bool,
    Bytes,
    Address,
    Timestamp,
    Chain_id,
    Bls12_381_fr,
    Bls12_381_g1,
    Bls12_381_g2,
    Key,
    Key_hash,
    Signature,
    // Containers
    List,
    Set,
    Some,
    None,
    Pair,
    Map,
    Big_map,
    // Lambda,
    Ticket,
    Sapling_state,
    Record,
    Variant,
};

export default Literals;
