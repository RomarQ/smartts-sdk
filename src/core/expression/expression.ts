import { proxy } from '../../misc/proxy';
import { LineInfo } from '../../misc/utils';
import { IExpression } from '../../typings/expression';
import { ILiteral } from '../../typings/literal';
import { IType } from '../../typings/type';

export class Expression implements IExpression {
    _isExpression = true as const;
    args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string | ILiteral<unknown>)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return Expression.getAttr(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    static getAttr = (from: IExpression, attr: string, line = new LineInfo()) =>
        proxy(new Expression('attr', from, `"${attr}"`, line), Expression.proxyHandler);

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}
