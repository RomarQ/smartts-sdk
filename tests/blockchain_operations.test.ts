import { Address, ChainID, Nat, Timestamp } from '../src/core/literal';
import { GetChainID, Contract, EntryPoint, GetCurrentTime, GetLevel, GetSender, GetSource } from '../src/core';
import { ContractStorage } from '../src/core/expression';
import { SetValue } from '../src/core/command';
import SmartML from '../src/smartml';

const verifyMichelsonOutput = (contract: string) => {
    const michelson = SmartML.compileContract(contract);

    // No errors expected
    expect(JSON.stringify(michelson).includes('ERROR')).toBeFalsy();
    // Check snapshot
    expect(michelson).toMatchSnapshot();
};

describe('Test compilation of blockchain operations', () => {
    it('Get sender', () => {
        const contract = new Contract({
            initialStorage: Address('tz1'),
            entries: [new EntryPoint('ep1').code(() => [SetValue(ContractStorage(), GetSender())])],
        }).toString();

        expect(contract).toMatchSnapshot();
        verifyMichelsonOutput(contract);
    });
    it('Get source', () => {
        const contract = new Contract({
            initialStorage: Address('tz1'),
            entries: [new EntryPoint('ep1').code(() => [SetValue(ContractStorage(), GetSource())])],
        }).toString();

        expect(contract).toMatchSnapshot();
        verifyMichelsonOutput(contract);
    });
    it('Get head level', () => {
        const contract = new Contract({
            initialStorage: Nat(1),
            entries: [new EntryPoint('ep1').code(() => [SetValue(ContractStorage(), GetLevel())])],
        }).toString();

        expect(contract).toMatchSnapshot();
        verifyMichelsonOutput(contract);
    });
    it('Get head timestamp', () => {
        const contract = new Contract({
            initialStorage: Timestamp(0),
            entries: [new EntryPoint('ep1').code(() => [SetValue(ContractStorage(), GetCurrentTime())])],
        }).toString();

        expect(contract).toMatchSnapshot();
        verifyMichelsonOutput(contract);
    });
    it('Get chain identifier', () => {
        const contract = new Contract({
            initialStorage: ChainID('0x00'),
            entries: [new EntryPoint('ep1').code(() => [SetValue(ContractStorage(), GetChainID())])],
        }).toString();

        expect(contract).toMatchSnapshot();
        verifyMichelsonOutput(contract);
    });
});
