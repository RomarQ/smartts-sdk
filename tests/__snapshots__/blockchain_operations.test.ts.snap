// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test compilation of blockchain operations Get chain identifier 1`] = `
"
        (
            template_id (static_id 0 (\\"blockchain_operations.test.ts\\" 54))
            storage (literal (chain_id_cst 0x00) (\\"blockchain_operations.test.ts\\" 55))
            storage_type (\\"chain_id\\")
            messages ((ep1 True False False False (\\"blockchain_operations.test.ts\\" 56) ((set (data) (chain_id) (\\"blockchain_operations.test.ts\\" 56)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"blockchain_operations.test.ts\\" 54))
        )
        "
`;

exports[`Test compilation of blockchain operations Get chain identifier 2`] = `
Array [
  Object {
    "args": Array [
      Object {
        "prim": "chain_id",
      },
    ],
    "prim": "storage",
  },
  Object {
    "args": Array [
      Object {
        "annots": Array [
          "%ep1",
        ],
        "prim": "unit",
      },
    ],
    "prim": "parameter",
  },
  Object {
    "args": Array [
      Array [
        Object {
          "prim": "CAR",
        },
        Object {
          "prim": "CHAIN_ID",
        },
        Object {
          "prim": "SWAP",
        },
        Object {
          "prim": "DROP",
        },
        Object {
          "args": Array [
            Object {
              "prim": "operation",
            },
          ],
          "prim": "NIL",
        },
        Object {
          "prim": "PAIR",
        },
      ],
    ],
    "prim": "code",
  },
]
`;

exports[`Test compilation of blockchain operations Get head level 1`] = `
"
        (
            template_id (static_id 0 (\\"blockchain_operations.test.ts\\" 36))
            storage (literal (nat 1) (\\"blockchain_operations.test.ts\\" 37))
            storage_type (\\"nat\\")
            messages ((ep1 True False False False (\\"blockchain_operations.test.ts\\" 38) ((set (data) (level) (\\"blockchain_operations.test.ts\\" 38)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"blockchain_operations.test.ts\\" 36))
        )
        "
`;

exports[`Test compilation of blockchain operations Get head level 2`] = `
Array [
  Object {
    "args": Array [
      Object {
        "prim": "nat",
      },
    ],
    "prim": "storage",
  },
  Object {
    "args": Array [
      Object {
        "annots": Array [
          "%ep1",
        ],
        "prim": "unit",
      },
    ],
    "prim": "parameter",
  },
  Object {
    "args": Array [
      Array [
        Object {
          "prim": "CAR",
        },
        Object {
          "prim": "LEVEL",
        },
        Object {
          "prim": "SWAP",
        },
        Object {
          "prim": "DROP",
        },
        Object {
          "args": Array [
            Object {
              "prim": "operation",
            },
          ],
          "prim": "NIL",
        },
        Object {
          "prim": "PAIR",
        },
      ],
    ],
    "prim": "code",
  },
]
`;

exports[`Test compilation of blockchain operations Get head timestamp 1`] = `
"
        (
            template_id (static_id 0 (\\"blockchain_operations.test.ts\\" 45))
            storage (literal (timestamp 0) (\\"blockchain_operations.test.ts\\" 46))
            storage_type (\\"timestamp\\")
            messages ((ep1 True False False False (\\"blockchain_operations.test.ts\\" 47) ((set (data) (now) (\\"blockchain_operations.test.ts\\" 47)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"blockchain_operations.test.ts\\" 45))
        )
        "
`;

exports[`Test compilation of blockchain operations Get head timestamp 2`] = `
Array [
  Object {
    "args": Array [
      Object {
        "prim": "timestamp",
      },
    ],
    "prim": "storage",
  },
  Object {
    "args": Array [
      Object {
        "annots": Array [
          "%ep1",
        ],
        "prim": "unit",
      },
    ],
    "prim": "parameter",
  },
  Object {
    "args": Array [
      Array [
        Object {
          "prim": "DROP",
        },
        Object {
          "prim": "NOW",
        },
        Object {
          "args": Array [
            Object {
              "prim": "operation",
            },
          ],
          "prim": "NIL",
        },
        Object {
          "prim": "PAIR",
        },
      ],
    ],
    "prim": "code",
  },
]
`;

exports[`Test compilation of blockchain operations Get sender 1`] = `
"
        (
            template_id (static_id 0 (\\"blockchain_operations.test.ts\\" 18))
            storage (literal (address tz1) (\\"blockchain_operations.test.ts\\" 19))
            storage_type (\\"address\\")
            messages ((ep1 True False False False (\\"blockchain_operations.test.ts\\" 20) ((set (data) (sender) (\\"blockchain_operations.test.ts\\" 20)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"blockchain_operations.test.ts\\" 18))
        )
        "
`;

exports[`Test compilation of blockchain operations Get sender 2`] = `
Array [
  Object {
    "args": Array [
      Object {
        "prim": "address",
      },
    ],
    "prim": "storage",
  },
  Object {
    "args": Array [
      Object {
        "annots": Array [
          "%ep1",
        ],
        "prim": "unit",
      },
    ],
    "prim": "parameter",
  },
  Object {
    "args": Array [
      Array [
        Object {
          "prim": "DROP",
        },
        Object {
          "prim": "SENDER",
        },
        Object {
          "args": Array [
            Object {
              "prim": "operation",
            },
          ],
          "prim": "NIL",
        },
        Object {
          "prim": "PAIR",
        },
      ],
    ],
    "prim": "code",
  },
]
`;

exports[`Test compilation of blockchain operations Get source 1`] = `
"
        (
            template_id (static_id 0 (\\"blockchain_operations.test.ts\\" 27))
            storage (literal (address tz1) (\\"blockchain_operations.test.ts\\" 28))
            storage_type (\\"address\\")
            messages ((ep1 True False False False (\\"blockchain_operations.test.ts\\" 29) ((set (data) (source) (\\"blockchain_operations.test.ts\\" 29)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"blockchain_operations.test.ts\\" 27))
        )
        "
`;

exports[`Test compilation of blockchain operations Get source 2`] = `
Array [
  Object {
    "args": Array [
      Object {
        "prim": "address",
      },
    ],
    "prim": "storage",
  },
  Object {
    "args": Array [
      Object {
        "annots": Array [
          "%ep1",
        ],
        "prim": "unit",
      },
    ],
    "prim": "parameter",
  },
  Object {
    "args": Array [
      Array [
        Object {
          "prim": "DROP",
        },
        Object {
          "prim": "SOURCE",
        },
        Object {
          "args": Array [
            Object {
              "prim": "operation",
            },
          ],
          "prim": "NIL",
        },
        Object {
          "prim": "PAIR",
        },
      ],
    ],
    "prim": "code",
  },
]
`;
